File: 2121-find-if-path-exists-in-graph/2121-find-if-path-exists-in-graph.py
class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        graph = defaultdict(list)
        for a, b in edges:
            graph[a].append(b)
            graph[b].append(a)
        q = deque([source])
        seen = {source}
        while q:
            node = q.popleft()
            if node == destination:
                return True
            for nb in graph[node]:
                if nb not in seen:
                    q.append(nb)
                    seen.add(nb)
        return False

File: 2121-find-if-path-exists-in-graph/2121-find-if-path-exists-in-graph.java
class Solution {
    public boolean validPath(int n, int[][] edges, int source, int destination) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        for (int[] edge : edges) {
            int u = edge[0];
            int v = edge[1];
            graph.computeIfAbsent(u, k -> new ArrayList<>()).add(v);
            graph.computeIfAbsent(v, k -> new ArrayList<>()).add(u);
        }
        Set<Integer> seen = new HashSet<>();
        Stack<Integer> q = new Stack<>();
        q.add(source);
        seen.add(source);
        while (!q.isEmpty()) {
            int node = q.pop();
            List<Integer> nbs = graph.get(node);
            if (node == destination) {
                return true;
            }
            for (int nb : nbs) {
                if (!seen.contains(nb)) {
                    q.add(nb);
                    seen.add(nb);
                }
            }
        }
        return false;
    }
}